src/ros_robot_localization_listener.cpp

```diff
@@ -0,0 +1,548 @@
+/*
+ * Copyright (c) 2016, TNO IVS Helmond.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following
+ * disclaimer in the documentation and/or other materials provided
+ * with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ * contributors may be used to endorse or promote products derived
+ * from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <exception>
+#include <functional>
+#include <string>
+#include <vector>
+
+#include <rclcpp/qos.hpp>
+
+#include "robot_localization/ros_robot_localization_listener.hpp"
+#include "robot_localization/ros_filter_utilities.hpp"
+
+#include <Eigen/Dense>
+#include <tf2/LinearMath/Matrix3x3.h>
+#include <tf2/LinearMath/Quaternion.h>
+#include <tf2/time.h>
+#include <tf2_eigen/tf2_eigen.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#include <yaml-cpp/yaml.h>
+
+#define THROTTLE(clock, duration, thing) do { \
+  static rclcpp::Time _last_output_time##__LINE__(0, 0, (clock)->get_clock_type()); \
+  auto _now = (clock)->now();                                           \
+  if (_now - _last_output_time##__LINE__ > (duration)) {                          \
+    _last_output_time##__LINE__ = _now; \
+    thing; \
+  } \
+} while(0)
+
+namespace robot_localization
+{
+FilterType filterTypeFromString(const std::string& filter_type_str)
+{
+  if ( filter_type_str == "ekf" )
+  {
+    return FilterTypes::EKF;
+  }
+  else if ( filter_type_str == "ukf" )
+  {
+    return FilterTypes::UKF;
+  }
+  else
+  {
+    return FilterTypes::NotDefined;
+  }
+}
+
+RosRobotLocalizationListener::RosRobotLocalizationListener(
+  rclcpp::Node::SharedPtr node=rclcpp::Node::make_shared("robot_localization")):
```

Comment:
I'm super duper not a fan of having nodes being implicitly created given how much overhead they create in ROS2. Is there a reason we need this default argument?

Reply:
I am not sure whether it is needed. The original [constructor](https://github.com/cra-ros-pkg/robot_localization/blob/melodic-devel/src/ros_robot_localization_listener.cpp#L69) in Melodic constructs a `NodeHandle` with `robot_localization`, and I don't know if anyone using it assumes that namespace in the topic names. So I just kept it. I'm constructing it with a different name in test and using that name in the test launch. The only other use case in the code is in `ros_robot_localization_listener.cpp`, and that uses a different name too.

It needs to be a valid node pointer, because `odom_sub_` and `accel_sub_` are constructed using it. I don't know whether it fails safely if the pointer is invalid.

I removed the default argument in 926a4e4.

Reply:
Makes sense. Nodehandles in ROS1 (to my understanding, which I'm opening myself up to being completely wrong) had trivial overhead. I'd like make'm to grab a param and trash them. ROS2 Nodes are a different beast creating their own DDS participants and coming with alot of online networking traffic. 

But it sounds like you just removed the default argument, so that resolves my concern.

Reply:
Interesting. The Melodic version did actually have a second node in the constructor doing exactly what you just said, grabbing a few parameters, heh... I guess that isn't weird.